### PR TITLE
Add `to_dgl` and `from_dgl` conversions

### PR DESCRIPTION
**Pull request by hbenedek**
Issue https://github.com/pyg-team/pytorch_geometric/issues/6979

## Description

I add two functions `from_dgl(dgl.Graph) -> data: Union[Data, HeteroData]`and
 `to_dgl(Union[Data, HeteroData]) -> dgl.Graph` in `utils/convert.py`. Both function in both direction performs the same three operations, setting the node and edge attributes and copies the edges of the underlying graph(s).

### CODE REVIEW

File: torch_geometric/utils/convert.py [hunk 1]
```diff
@@ -455,3 +455,90 @@ def from_cugraph(g: Any) -> Tuple[Tensor, Optional[Tensor]]:
...
+            The data object.
+    """
+    import dgl
+
+    from torch_geometric.data import Data, HeteroData
+
+    if isinstance(data, Data):
+        g = dgl.graph((data.edge_index[0], data.edge_index[1]))
+        for attr in data.node_attrs():
+            g.ndata[attr] = getattr(data, attr)
```
Comment (EdisonLeeeee) [inline torch_geometric/utils/convert.py:478]:
```suggestion
            g.ndata[attr] = data.get(attr)
```

File: torch_geometric/utils/convert.py [hunk 2]
```diff
@@ -455,3 +455,90 @@ def from_cugraph(g: Any) -> Tuple[Tensor, Optional[Tensor]]:
...
+    Args:
+        data (torch_geometric.data.Data or torch_geometric.data.HeteroData):
+            The data object.
+    """
+    import dgl
+
+    from torch_geometric.data import Data, HeteroData
+
+    if isinstance(data, Data):
+        g = dgl.graph((data.edge_index[0], data.edge_index[1]))
```
Comment (EdisonLeeeee) [inline torch_geometric/utils/convert.py:476]: We should also take `adj_t` into consideration.

File: torch_geometric/utils/convert.py [hunk 3]
```diff
@@ -455,3 +455,90 @@ def from_cugraph(g: Any) -> Tuple[Tensor, Optional[Tensor]]:
...
+        for attr in data.node_attrs():
+            g.ndata[attr] = getattr(data, attr)
+        for attr in data.edge_attrs():
+            if attr == "edge_index":
+                continue
+            g.edata[attr] = getattr(data, attr)
```
Comment (EdisonLeeeee) [inline torch_geometric/utils/convert.py:482]:
```suggestion
            g.edata[attr] = data.get(attr)
```

File: torch_geometric/utils/convert.py [hunk 4]
```diff
@@ -455,3 +455,90 @@ def from_cugraph(g: Any) -> Tuple[Tensor, Optional[Tensor]]:
...
+
+def from_dgl(
+    g: Any,
+) -> Union["torch_geometric.data.Data", "torch_geometric.data.HeteroData"]:
+    r"""Converts a :obj:`dgl` graph object to a
+    :class:`torch_geometric.data.Data` or
+    :class:`torch_geometric.data.HeteroData` instance.
+
+    Args:
+        g (dgl.DGLGraph): The :obj:`dgl` graph object.
```
Comment (EdisonLeeeee) [inline torch_geometric/utils/convert.py:514]: Add an example here :)

**Review by EdisonLeeeee — COMMENTED** [review]
Thanks for the PR! Please also includes some test cases for these two functions.

File: torch_geometric/utils/convert.py (hunk 2 above)
Comment (hbenedek) [reply torch_geometric/utils/convert.py:499]: @EdisonLeeeee Thank you for your suggestions. Can you elaborate a bit on this one? Not sure if I understand what you mean by considering `adj_t`.

File: torch_geometric/utils/convert.py [hunk 5]
```diff
@@ -455,3 +455,90 @@ def from_cugraph(g: Any) -> Tuple[Tensor, Optional[Tensor]]:
...
+    data: Union["torch_geometric.data.Data", "torch_geometric.data.HeteroData"]
+) -> Any:
+    r"""Converts a :class:`torch_geometric.data.Data` or
+    :class:`torch_geometric.data.HeteroData` instance to a :obj:`dgl` graph
+    object.
+
+    Args:
+        data (torch_geometric.data.Data or torch_geometric.data.HeteroData):
+            The data object.
+    """
```
[the remark this answers is not in the document]
Comment (hbenedek) [reply torch_geometric/utils/convert.py:493]: I added two examples in both cases, one for HeteroData and one for Data, should I also create tests in `test/utils/test_convert.py`?

File: torch_geometric/utils/convert.py (hunk 2 above)
Comment (EdisonLeeeee) [reply torch_geometric/utils/convert.py:499]: As there may be cases where data only contains the transposed sparse adjacency matrix `adj_t`, we need to take this into consideration as well. For example, `data` was transformed by `ToSparseTensor`.

File: torch_geometric/utils/convert.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (EdisonLeeeee) [reply torch_geometric/utils/convert.py:493]: Yep, that would be better :)

**Comment by hbenedek** [comment]
@EdisonLeeeee I added the tests, but for some reason, codecov complains about not seeing them. I ran `pytest --cov` and the coverage of `torch_geometric/utils/convert.py` is only 9%. Am I doing something wrong, or just the external libraries causing this issue?

**Comment by rusty1s** [comment]
@hbenedek Don't worry about it. It's good that we have a test but it's not called by CI due to missing dependency.

**Review by rusty1s — APPROVED** [review]

**Comment by rusty1s** [comment]
Great PR, thanks for doing that :)

**Comment by EdisonLeeeee** [comment]
@hbenedek thank you, too :D

**Pull request merged**